### PR TITLE
Add a new method BulkFindEntities

### PR DESCRIPTION
**Pull request by Spaceenter**

### CODE REVIEW

File: proto/recon.proto [hunk 1]
```diff
@@ -109,4 +109,23 @@ message ResolveIdsResponse {
...
+    // Optional.
+    string type = 2;
+  }
+  repeated Entity entities = 1;
+}
+
+message BulkFindEntitiesResponse {
+  message Entity {
+    string description = 1;
+    // Optional.
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply proto/recon.proto:126]: I think there will be cases when users only know the name, but not the type? This is especially so given that the type is "Data Commons type"?

File: proto/mixer.proto [hunk 2]
```diff
@@ -629,4 +629,19 @@ service Mixer {
...
     };
   }
+
+
+  // TODO(spaceenter): Add a non-bulk method FindEntities().
+
+  // Find entities from descriptions, with optional filters on types.
+  rpc BulkFindEntities(BulkFindEntitiesRequest) returns (BulkFindEntitiesResponse) {
+    option (google.api.http) = {
+      get : "/v1/bulk/find/entities"
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply proto/mixer.proto:639]: OK, so maybe for the "Bulk" method we only do POST, and don't do GET.
Maybe I could add a GET for the non-bulk method. 
Changed.

File: internal/util/util.go [hunk 3]
```diff
@@ -482,3 +492,39 @@ func StringListIntersection(list [][]string) []string {
 
 	return res
 }
+
+// MapsClient gets the client for Maps API.
+func MapsClient(ctx context.Context) (*maps.Client, error) {
+	var projectID string
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/util/util.go:498]: Thanks for letting me know about this! Changed.

File: internal/util/util.go [hunk 4]
```diff
@@ -482,3 +492,39 @@ func StringListIntersection(list [][]string) []string {
...
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	return maps.NewClient(maps.WithAPIKey(string(secret.Payload.Data)))
+}
+
+// StringSetToSlice is a helper to convert a string set to a string slice.
+func StringSetToSlice(s map[string]struct{}) []string {
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/util/util.go:524]: Sorry the link doesn't point to a function?

File: internal/server/recon/golden/bulk_find_entities_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		}{
+			{
+				&pb.BulkFindEntitiesRequest{
+					Entities: []*pb.BulkFindEntitiesRequest_Entity{
+						{Description: "Santa Clara County"},
+						{Description: "Santa Clara County", Type: "County"},
+						{Description: "Cambridge, MA"},
+						{Description: "mountain view"},
+						{Description: "Mountain View, CA"},
+						// Typo on purpose.
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/golden/bulk_find_entities_test.go:49]: Thanks to Maps API!

File: internal/server/recon/golden/bulk_find_entities/result.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{
+  "entities": [
+    {
+      "description": "Cambridge, MA",
+      "dcids": [
+        "geoId/2511000"
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/golden/bulk_find_entities/result.json:6]: "Cambridge" seems to be a corner case that local call to Maps API returns different results from GCB call to Maps API... I'll try to find some other place with multiple DCIDs returned...

[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/golden/bulk_find_entities/result.json:6]: Used Newton City.

File: internal/server/recon/find_entities.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+)
+
+const (
+	maxNumEntitiesPerRequest  = 5000
+	maxMapsAPICallsInParallel = 25
+)
+
+type entityInfo struct {
+	description string
+	entityType  string
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/find_entities.go:38]: done

File: internal/server/recon/find_entities.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+			"exceeded max number of entities per request (%d): %d",
+			maxNumEntitiesPerRequest, l)
+	}
+
+	// Load input.
+	entityInfoSet := map[entityInfo]struct{}{}
+	for _, entity := range in.GetEntities() {
+		description := entity.GetDescription()
+		if description == "" {
+			return nil, fmt.Errorf("empty entity description")
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/find_entities.go:61]: changed

File: internal/server/recon/find_entities.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+func resolvePlaceIDs(
+	ctx context.Context,
+	mapsClient *maps.Client,
+	entityInfoSet map[entityInfo]struct{},
+) (
+	map[entityInfo][]string, /* entityInfo -> [place ID] */
+	map[string]struct{}, /* [place ID] for all entities */
+	error) {
```
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/find_entities.go:132]: done

File: internal/server/recon/golden/bulk_find_entities/result.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/server/recon/golden/bulk_find_entities/result.json]: "Newton City" behaves similarly as "Cambridge".... Locally I get 3 cities from the Maps API, however, on GCB it only returns one...

I'll skip these corner cases from the test now, I want to see if we call Maps API from autopush/staging/prod, if it works OK, then we'll know if it's an issue of GCB...

File: internal/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shifucun) [reply internal/util/util.go:512]: I mean KeysToSlice() func in this file. But that caller may need to change. Up to you if you think it's worth to consolidate.

File: proto/recon.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (shifucun) [reply proto/recon.proto:126]: Oh, I see, this is the type from the request. I wonder how would user uses the response. They need to match the description and/or type?

**Review by shifucun — APPROVED** [review]

File: proto/recon.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (Spaceenter) [reply proto/recon.proto:126]: Yes

File: internal/util/util.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Spaceenter) [reply internal/util/util.go:512]: Maybe keep for now, as I also don't need to sort here...
